ecs_client_test.go

```diff
@@ -39,49 +60,205 @@ func TestECSClientRegisterAndDeregisterTaskDefinition(t *testing.T) {
 		HTTPClient: hc,
 	})
 	require.NoError(t, err)
+	require.NotNil(t, c)
 
-	t.Run("RegisterFailsWithInvalidInput", func(t *testing.T) {
-		out, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{})
-		assert.Error(t, err)
-		assert.Zero(t, out)
-	})
-	t.Run("DeregisterFailsWithInvalidInput", func(t *testing.T) {
-		out, err := c.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{})
-		assert.Error(t, err)
-		assert.Zero(t, out)
+	registerOut, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+		ContainerDefinitions: []*ecs.ContainerDefinition{
+			{
+				Command: []*string{aws.String("echo"), aws.String("foo")},
+				Image:   aws.String("ubuntu"),
+				Name:    aws.String("print_foo"),
+			},
+		},
+		Cpu:    aws.String("128"),
+		Memory: aws.String("4"),
+		Family: aws.String(os.Getenv("AWS_ECS_TASK_DEFINITION_PREFIX") + "bar"),
 	})
-	t.Run("RegisterAndDeregisterSucceeds", func(t *testing.T) {
-		out, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
-			ContainerDefinitions: []*ecs.ContainerDefinition{
-				{
-					Command: []*string{aws.String("echo"), aws.String("hello")},
-					Image:   aws.String("ubuntu"),
-					Name:    aws.String("hello_world"),
+	require.NoError(t, err)
+	require.NotZero(t, registerOut)
+	require.NotZero(t, registerOut.TaskDefinition)
+
+	defer func() {
+		cleanupTaskDefinition(ctx, t, c, registerOut)
+		c.Close(ctx)
+	}()
+
+	for tName, tCase := range map[string]func(context.Context, *testing.T, *BasicECSClient){
+		"RegisterFailsWithInvalidInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"DeregisterFailsWithInvalidInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"RegisterAndDeregisterSucceeds": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+				ContainerDefinitions: []*ecs.ContainerDefinition{
+					{
+						Command: []*string{aws.String("echo"), aws.String("hello")},
+						Image:   aws.String("ubuntu"),
+						Name:    aws.String("hello_world"),
+					},
 				},
-			},
-			Cpu:    aws.String("128"),
-			Memory: aws.String("4"),
-			Family: aws.String("foo"),
-		})
-		require.NoError(t, err)
-		require.NotZero(t, out)
+				Cpu:    aws.String("128"),
+				Memory: aws.String("4"),
+				Family: aws.String(os.Getenv("AWS_ECS_TASK_DEFINITION_PREFIX") + "foo"),
+			})
+			require.NoError(t, err)
+			require.NotZero(t, out)
+
+			defer func() {
```

Comment:
Isn't this the same as `cleanupTaskDefinition`?

Reply:
Yes, I didn't want to just call cleanupTaskDefinition because it doesn't make it super obvious that we are also testing de-register, but I am refactoring it so that it's not in a `defer`.